// @antv/g2plot/esm/plots/stacked-bar/layer.js
import { __assign, __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import BaseBarLayer from '../bar/layer';
import StackBarLabel from './component/label';
var StackedBarLayer = /** @class */ (function (_super) {
    __extends(StackedBarLayer, _super);
    function StackedBarLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'stackedBar';
        return _this;
    }
    StackedBarLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            xAxis: {
                visible: true,
                autoRotateTitle: false,
                grid: {
                    visible: true,
                },
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: true,
                },
                label: {
                    visible: true,
                    autoRotate: true,
                    autoHide: true,
                },
                title: {
                    visible: true,
                    offset: 12,
                },
            },
            yAxis: {
                visible: true,
                autoRotateTitle: true,
                grid: {
                    visible: false,
                },
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                label: {
                    visible: true,
                    autoRotate: true,
                    autoHide: true,
                },
                title: {
                    visible: false,
                    offset: 12,
                },
            },
            legend: {
                visible: true,
                position: 'top-left',
                offsetY: 0,
            },
        });
    };
    StackedBarLayer.prototype.adjustBar = function (bar) {
        bar.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedBarLayer.prototype.renderLabel = function () {
        var scales = this.config.scales;
        var yField = this.options.yField;
        var scale = scales[yField];
        if (this.options.label && this.options.label.visible) {
            var label = new StackBarLabel(__assign({ view: this.view, plot: this, formatter: scale.formatter }, this.options.label));
            label.render();
        }
    };
    StackedBarLayer.prototype.geometryTooltip = function () {
        this.bar.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.bar.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.bar.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.bar.tooltip.fields = [this.options.xField, this.options.yField, this.options.stackField];
            }
        }
    };
    return StackedBarLayer;
}(BaseBarLayer));
export default StackedBarLayer;
registerPlotType('stackedBar', StackedBarLayer);
