// @antv/g2plot/esm/plots/gauge/geometry/shape/options.js
import { __assign } from "tslib";
/**
 * @author linhuiw
 * @description 存放仪表盘的配置项，比如指针的颜色、粗细等
 */
import { get } from '@antv/util';
var basicPointerStyleDark = {
    color: '#CFCFCF',
    circleColorTop: '#2E364B',
    circleColorBottom: '#EEEEEE',
    thickness: 4.5,
};
export var getOptions = function (name, theme, colors) {
    var basicPointerStyle = basicPointerStyleDark;
    var ringBackground = '#f0f0f0';
    var common = {
        color: get(colors, [0], '#F6445A'),
        thickness: 24,
        radius: 1,
        angle: 240,
        textPosition: '100%',
    };
    switch (name) {
        case 'standard':
            return {
                ringStyle: __assign(__assign({}, common), { background: ringBackground, axis: {
                        amount: 21,
                        offset: -35,
                        length: 5,
                        thickness: 2,
                        color: '#999',
                    } }),
                pointerStyle: __assign(__assign({}, basicPointerStyle), { radius: 0.5 }),
            };
        case 'meter':
            return {
                ringStyle: __assign(__assign({}, common), { background: ringBackground, axis: {
                        amount: 25,
                        offset: -35,
                        length: 2,
                        thickness: 1,
                        color: '#999',
                    } }),
                pointerStyle: __assign(__assign({}, basicPointerStyle), { radius: 0.5 }),
            };
        case 'fan':
            return {
                ringStyle: {
                    color: get(colors, [0], '#F6445A'),
                    background: ringBackground,
                    thickness: 70,
                    radius: 1,
                    angle: 120,
                    textPosition: '125%',
                    bottomRatio: 3.5,
                    axis: {
                        amount: 10,
                        offset: 5,
                        length: 3,
                        thickness: 3,
                        color: '#999',
                    },
                },
                pointerStyle: __assign(__assign({}, basicPointerStyle), { radius: 0.6 }),
            };
        default:
            break;
    }
};
