// @antv/g2plot/esm/plots/stacked-bar/component/label.js
import { __extends } from "tslib";
import { clone } from '@antv/util';
import BarLabel from '../../bar/component/label';
var DEFAULT_OFFSET = 8;
var StackBarLabel = /** @class */ (function (_super) {
    __extends(StackBarLabel, _super);
    function StackBarLabel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StackBarLabel.prototype.getDefaultOptions = function () {
        var theme = this.plot.theme;
        var labelStyle = theme.label.style;
        return {
            offsetX: DEFAULT_OFFSET,
            offsetY: 0,
            style: clone(labelStyle),
            adjustPosition: true,
        };
    };
    StackBarLabel.prototype.adjustLabel = function (label, shape) {
        if (this.options.adjustPosition) {
            var labelRange = label.getBBox();
            var shapeRange = this.getShapeBbox(shape);
            if (shapeRange.width <= labelRange.width) {
                // const xPosition = shapeRange.maxX + this.options.offsetX;
                label.attr('text', '');
            }
        }
    };
    return StackBarLabel;
}(BarLabel));
export default StackBarLabel;
