// @antv/g2plot/esm/base/view-layer.js
import { __extends, __spreadArrays } from "tslib";
import { deepMix, isEmpty, mapValues, get, isUndefined, each, assign, isFunction, mix, map, flatten, reduce, findIndex, isString, } from '@antv/util';
import { View, VIEW_LIFE_CIRCLE, registerComponentController, Gestrue } from '../dependents';
import TextDescription from '../components/description';
import { getLabelComponent } from '../components/label/base';
import { getComponent } from '../components/factory';
import BaseInteraction from '../interaction/index';
import { EVENT_MAP, onEvent } from '../util/event';
import PaddingController from './controller/padding';
import StateController from './controller/state';
import ThemeController from './controller/theme';
import Layer from './layer';
import { isTextUsable } from '../util/common';
registerComponentController('gesture', Gestrue);
var ViewLayer = /** @class */ (function (_super) {
    __extends(ViewLayer, _super);
    function ViewLayer(props) {
        var _this = _super.call(this, props) || this;
        _this.interactions = [];
        _this.labels = [];
        _this.options = _this.getOptions(props);
        _this.initialOptions = deepMix({}, _this.options);
        _this.paddingController = new PaddingController({
            plot: _this,
        });
        _this.stateController = new StateController({
            plot: _this,
        });
        _this.themeController = new ThemeController();
        return _this;
    }
    ViewLayer.getDefaultOptions = function () {
        return {
            renderer: 'canvas',
            title: {
                visible: false,
                alignTo: 'left',
                text: '',
            },
            description: {
                visible: false,
                text: '',
                alignTo: 'left',
            },
            padding: 'auto',
            legend: {
                visible: true,
                position: 'bottom-center',
            },
            tooltip: {
                visible: true,
                shared: true,
                showCrosshairs: true,
                crosshairs: {
                    type: 'x',
                },
                offset: 20,
            },
            xAxis: {
                visible: true,
                grid: {
                    visible: false,
                },
                line: {
                    visible: true,
                },
                tickLine: {
                    visible: true,
                },
                label: {
                    visible: true,
                    autoRotate: true,
                    autoHide: true,
                },
                title: {
                    visible: false,
                    offset: 12,
                },
            },
            yAxis: {
                visible: true,
                grid: {
                    visible: true,
                },
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                label: {
                    visible: true,
                    autoHide: true,
                    autoRotate: false,
                },
                title: {
                    autoRotate: true,
                    visible: false,
                    offset: 12,
                },
            },
            label: {
                visible: false,
            },
            interactions: [{ type: 'tooltip' }, { type: 'legend-active' }, { type: 'legend-filter' }],
            animation: false,
        };
    };
    ViewLayer.prototype.getOptions = function (props) {
        var curOptions = this.options || {};
        var options = _super.prototype.getOptions.call(this, props);
        // @ts-ignore
        var defaultOptions = this.constructor.getDefaultOptions(props);
        // interactions 需要合并去重下
        var interactions = reduce(flatten(map([options, defaultOptions, curOptions, props], function (src) { return get(src, 'interactions', []); })), function (result, cur) {
            var idx = findIndex(result, function (item) { return item.type === cur.type; });
            if (idx >= 0) {
                result.splice(idx, 1);
            }
            return __spreadArrays(result, [cur]);
        }, []);
        return deepMix({}, options, defaultOptions, curOptions, props, { interactions: interactions });
    };
    ViewLayer.prototype.beforeInit = function () {
        _super.prototype.beforeInit.call(this);
    };
    ViewLayer.prototype.init = function () {
        var _this = this;
        _super.prototype.init.call(this);
        this.theme = this.themeController.getTheme(this.options, this.type);
        this.config = {
            data: this.processData(this.options.data),
            scales: {},
            legends: {},
            tooltip: {
                showTitle: true,
            },
            axes: {},
            coordinate: { type: 'cartesian' },
            geometries: [],
            annotations: [],
            interactions: [],
            theme: this.theme,
            panelRange: {},
            animate: {},
            views: [],
        };
        this.paddingController.clear();
        this.drawTitle();
        this.drawDescription();
        // 有些interaction要调整配置项，所以顺序提前
        this.interaction();
        this.coord();
        this.scale();
        this.axis();
        this.tooltip();
        this.legend();
        this.addGeometry();
        this.annotation();
        this.animation();
        this.viewRange = this.getViewRange();
        var region = this.viewRangeToRegion(this.viewRange);
        this.view = new View({
            parent: null,
            canvas: this.canvas,
            backgroundGroup: this.container.addGroup(),
            middleGroup: this.container.addGroup(),
            foregroundGroup: this.container.addGroup(),
            padding: this.paddingController.getPadding(),
            theme: this.theme,
            options: this.config,
            region: region,
        });
        this.applyInteractions();
        this.view.on(VIEW_LIFE_CIRCLE.AFTER_RENDER, function () {
            _this.afterRender();
        });
    };
    ViewLayer.prototype.afterInit = function () {
        _super.prototype.afterInit.call(this);
        if (!this.view || this.view.destroyed) {
            return;
        }
        if (this.options.padding !== 'auto') {
            this.parseEvents();
        }
    };
    ViewLayer.prototype.afterRender = function () {
        if (!this.view || this.view.destroyed) {
            return;
        }
        var options = this.options;
        var padding = options.padding ? options.padding : this.config.theme.padding;
        /** defaultState */
        if (options.defaultState && padding !== 'auto') {
            this.stateController.defaultStates(options.defaultState);
        }
        /** autopadding */
        if (padding === 'auto') {
            this.paddingController.processAutoPadding();
        }
        if (options.tooltip && options.tooltip.customContent && options.padding !== 'auto') {
            this.customTooltip();
        }
    };
    /** 完整生命周期渲染 */
    ViewLayer.prototype.render = function () {
        _super.prototype.render.call(this);
        var data = this.options.data;
        if (!isEmpty(data)) {
            this.view.render();
        }
    };
    /** 销毁 */
    ViewLayer.prototype.destroy = function () {
        this.doDestroy();
        _super.prototype.destroy.call(this);
    };
    /** 更新配置项 */
    ViewLayer.prototype.updateConfig = function (cfg) {
        this.doDestroy();
        if (!cfg.padding && this.initialOptions.padding && this.initialOptions.padding === 'auto') {
            cfg.padding = 'auto';
        }
        this.options = this.getOptions(cfg);
        this.processOptions(this.options);
    };
    ViewLayer.prototype.changeData = function (data) {
        this.options.data = this.processData(data);
        this.view.changeData(this.options.data);
    };
    // plot 不断销毁重建，需要一个api获取最新的plot
    ViewLayer.prototype.getPlot = function () {
        return this.view;
    };
    /**
     * 获取已渲染的数据标签组件
     */
    ViewLayer.prototype.getLabels = function () {
        return this.labels;
    };
    // 获取对应的G2 Theme
    ViewLayer.prototype.getTheme = function () {
        if (!this.theme) {
            return this.themeController.getTheme(this.options, this.type);
        }
        return this.theme;
    };
    ViewLayer.prototype.getResponsiveTheme = function () {
        return this.themeController.getResponsiveTheme(this.type);
    };
    // 获取对应的Plot Theme
    ViewLayer.prototype.getPlotTheme = function () {
        return this.themeController.getPlotTheme(this.options, this.type);
    };
    ViewLayer.prototype.getInteractions = function () {
        return this.interactions;
    };
    // 绑定一个外部的stateManager
    ViewLayer.prototype.bindStateManager = function (stateManager, cfg) {
        this.stateController.bindStateManager(stateManager, cfg);
    };
    // 响应状态量更新的快捷方法
    ViewLayer.prototype.setActive = function (condition, style) {
        this.stateController.setState({ type: 'active', condition: condition, style: style });
    };
    ViewLayer.prototype.setSelected = function (condition, style) {
        this.stateController.setState({ type: 'selected', condition: condition, style: style });
    };
    ViewLayer.prototype.setDisable = function (condition, style) {
        this.stateController.setState({ type: 'disable', condition: condition, style: style });
    };
    ViewLayer.prototype.setDefault = function (condition, style) {
        this.stateController.setState({ type: 'default', condition: condition, style: style });
    };
    // 获取 ViewLayer 的数据项
    ViewLayer.prototype.getData = function (start, end) {
        return this.processData((this.options.data || []).slice(start, end));
    };
    ViewLayer.prototype.processData = function (data) {
        return data;
    };
    ViewLayer.prototype.scale = function () {
        /** scale meta配置 */
        // 1. this.config.scales中已有子图形在处理xAxis/yAxis是写入的xField/yField对应的scale信息，这里再检查用户设置的meta，将meta信息合并到默认的scale中
        // 2. 同时xAxis/yAxis中的type优先级更高，覆盖meta中的type配置
        var scaleTypes = mapValues(this.config.scales, function (scaleConfig) {
            var type = scaleConfig.type;
            return type ? { type: type } : {};
        });
        var scales = deepMix({}, this.config.scales, this.options.meta || {}, scaleTypes);
        this.setConfig('scales', scales);
    };
    ViewLayer.prototype.axis = function () {
        var xAxis_parser = getComponent('axis', {
            plot: this,
            dim: 'x',
        });
        var yAxis_parser = getComponent('axis', {
            plot: this,
            dim: 'y',
        });
        var axesConfig = {};
        axesConfig[this.options.xField] = xAxis_parser;
        axesConfig[this.options.yField] = yAxis_parser;
        /** 存储坐标轴配置项到config */
        this.setConfig('axes', axesConfig);
    };
    ViewLayer.prototype.tooltip = function () {
        if (this.options.tooltip.visible === false) {
            this.setConfig('tooltip', false);
            return;
        }
        var tooltipOptions = get(this.options, 'tooltip');
        if (tooltipOptions.customContent && tooltipOptions.customContent.container) {
            tooltipOptions.container = tooltipOptions.customContent.container;
        }
        this.setConfig('tooltip', deepMix({}, tooltipOptions));
        deepMix(this.config.theme.tooltip, this.options.tooltip.domStyles);
    };
    ViewLayer.prototype.customTooltip = function () {
        var customContentCfg = this.options.tooltip.customContent;
        var container;
        if (customContentCfg.container) {
            container = isString(customContentCfg.container)
                ? document.getElementById(customContentCfg.container)
                : customContentCfg.container;
        }
        this.view.on('tooltip:show', function () {
            if (!customContentCfg.container) {
                container = document.getElementsByClassName('g2-tooltip')[0];
            }
        });
        this.view.hideTooltip();
        this.view.on('tooltip:change', function (ev) {
            if (container) {
                customContentCfg.callback(container, ev);
            }
        });
    };
    ViewLayer.prototype.getLegendPosition = function (position) {
        var positionList = position.split('-');
        // G2 4.0 兼容 XXX-center 到 XXX 的场景
        if (positionList && positionList.length > 1 && positionList[1] === 'center') {
            return positionList[0];
        }
        return position;
    };
    ViewLayer.prototype.legend = function () {
        if (this.options.legend.visible === false) {
            this.setConfig('legends', false);
            return;
        }
        var flipOption = get(this.options, 'legend.flipPage');
        var clickable = get(this.options, 'legend.clickable');
        this.setConfig('legends', {
            position: this.getLegendPosition(get(this.options, 'legend.position')),
            formatter: get(this.options, 'legend.formatter'),
            offsetX: get(this.options, 'legend.offsetX'),
            offsetY: get(this.options, 'legend.offsetY'),
            clickable: isUndefined(clickable) ? true : clickable,
            // wordSpacing: get(this.options, 'legend.wordSpacing'),
            flipPage: flipOption,
            marker: get(this.options, 'legend.marker'),
            title: get(this.options, 'legend.title'),
        });
    };
    ViewLayer.prototype.annotation = function () {
        var _this = this;
        var config = [];
        if (this.config.coordinate.type === 'cartesian' && this.options.guideLine) {
            each(this.options.guideLine, function (line) {
                var guideLine = getComponent('guideLine', {
                    plot: _this,
                    cfg: line,
                });
                config.push(guideLine);
            });
        }
        this.setConfig('annotations', config);
    };
    ViewLayer.prototype.interaction = function () {
        var _this = this;
        var _a = this.options.interactions, interactions = _a === void 0 ? [] : _a;
        each(interactions, function (interaction) {
            var type = interaction.type;
            if (type === 'slider' || type === 'scrollbar') {
                var axisConfig = {
                    label: {
                        autoHide: true,
                        autoRotate: false,
                    },
                };
                _this.options.xAxis = deepMix({}, _this.options.xAxis, axisConfig);
            }
            _this.setConfig('interaction', interaction);
        });
    };
    ViewLayer.prototype.animation = function () {
        if (this.options.animation === false) {
            this.setConfig('animate', false);
        }
    };
    ViewLayer.prototype.applyInteractions = function () {
        var _this = this;
        var _a = this.options.interactions, interactions = _a === void 0 ? [] : _a;
        if (this.interactions) {
            this.interactions.forEach(function (inst) {
                inst.destroy();
            });
        }
        this.interactions = [];
        interactions.forEach(function (interaction) {
            var Ctor = BaseInteraction.getInteraction(interaction.type, _this.type);
            if (Ctor) {
                var inst = new Ctor({ view: _this.view }, _this, Ctor.getInteractionRange(_this.layerBBox, interaction.cfg), interaction.cfg);
                _this.interactions.push(inst);
            }
        });
    };
    /** 设置G2 config，带有类型推导 */
    ViewLayer.prototype.setConfig = function (key, config) {
        if (key === 'geometry') {
            this.config.geometries.push(config);
            return;
        }
        if (key === 'interaction') {
            this.config.interactions.push(config);
            return;
        }
        if (config === false) {
            this.config[key] = false;
            return;
        }
        assign(this.config[key], config);
    };
    ViewLayer.prototype.parseEvents = function (eventParser) {
        var _this = this;
        var options = this.options;
        if (options.events) {
            _super.prototype.parseEvents.call(this, options.events);
            var eventmap_1 = eventParser ? eventParser.EVENT_MAP : EVENT_MAP;
            each(options.events, function (e, k) {
                if (isFunction(e)) {
                    var eventName = eventmap_1[k] || k;
                    var handler = e;
                    onEvent(_this, eventName, handler);
                }
            });
        }
    };
    ViewLayer.prototype.drawTitle = function () {
        var props = this.options;
        var range = this.layerBBox;
        if (this.title) {
            this.title.destroy();
            this.title = null;
        }
        if (isTextUsable(props.title)) {
            var width = this.width;
            var theme = this.config.theme;
            var title = new TextDescription({
                leftMargin: range.minX + theme.title.padding[3],
                rightMargin: range.maxX - theme.title.padding[1],
                topMargin: range.minY + theme.title.padding[0],
                text: props.title.text,
                style: mix(theme.title, props.title.style),
                wrapperWidth: width - theme.title.padding[3] - theme.title.padding[1],
                container: this.container.addGroup(),
                theme: theme,
                index: isTextUsable(props.description) ? 0 : 1,
                plot: this,
                alignTo: props.title.alignTo,
                name: 'title',
            });
            this.title = title;
            this.paddingController.registerPadding(title, 'outer');
        }
    };
    ViewLayer.prototype.drawDescription = function () {
        var props = this.options;
        var range = this.layerBBox;
        if (this.description) {
            this.description.destroy();
            this.description = null;
        }
        if (isTextUsable(props.description)) {
            var width = this.width;
            var theme = this.config.theme;
            var topMargin = 0;
            if (this.title) {
                var titleBBox = this.title.getBBox();
                topMargin += titleBBox.minY + titleBBox.height;
                topMargin += theme.description.padding[0];
            }
            else {
                // 无title的情况下使用title的上padding
                topMargin += range.minY + theme.title.padding[0];
            }
            var description = new TextDescription({
                leftMargin: range.minX + theme.description.padding[3],
                topMargin: topMargin,
                rightMargin: range.maxX - theme.title.padding[1],
                text: props.description.text,
                style: mix(theme.description, props.description.style),
                wrapperWidth: width - theme.description.padding[3] - theme.description.padding[1],
                container: this.container.addGroup(),
                theme: theme,
                index: 1,
                plot: this,
                alignTo: props.description.alignTo,
                name: 'description',
            });
            this.description = description;
            this.paddingController.registerPadding(description, 'outer');
        }
    };
    ViewLayer.prototype.doRenderLabel = function (geometry, label) {
        each(this.labels, function (item) {
            item.destroy();
        });
        this.labels = [];
        var config = {
            layer: this,
            container: geometry.labelsContainer,
            geometry: geometry,
            label: label,
        };
        var Ctor = getLabelComponent(label.type);
        if (Ctor) {
            var label_1 = new Ctor(config);
            label_1.render();
            this.labels.push(label_1);
        }
    };
    /** 抽取destroy和updateConfig共有代码为_destroy方法 */
    ViewLayer.prototype.doDestroy = function () {
        this.doDestroyInteractions();
        this.doDestroyLabels();
        /** 销毁g2.view实例 */
        if (!this.view.destroyed) {
            this.view.destroy();
        }
    };
    ViewLayer.prototype.doDestroyInteractions = function () {
        // 移除注册的 interactions
        if (this.interactions) {
            this.interactions.forEach(function (inst) {
                inst.destroy();
            });
        }
        this.interactions = [];
    };
    ViewLayer.prototype.doDestroyLabels = function () {
        // 移除各 geometry 的 label
        each(this.labels, function (label) {
            label.destroy();
        });
        this.labels = [];
    };
    ViewLayer.prototype.getViewRange = function () {
        var _this = this;
        // 有 Range 的 Interaction 参与 ViewMargin 计算
        var _a = this.options.interactions, interactions = _a === void 0 ? [] : _a;
        var layerBBox = this.layerBBox;
        interactions.forEach(function (interaction) {
            var Ctor = BaseInteraction.getInteraction(interaction.type, _this.type);
            var range = Ctor && Ctor.getInteractionRange(layerBBox, interaction.cfg);
            var position = '';
            if (range) {
                // 先只考虑 Range 靠边的情况
                if (range.maxY === layerBBox.maxY && range.minY > layerBBox.minY) {
                    // margin[2] += range.height;
                    position = 'bottom';
                }
                else if (range.maxX === layerBBox.maxX && range.minX > layerBBox.minX) {
                    // margin[1] += range.width;
                    position = 'right';
                }
                else if (range.minX === layerBBox.minX && range.maxX > layerBBox.maxX) {
                    // margin[3] += range.width;
                    position = 'left';
                }
                else if (range.minY === layerBBox.minY && range.maxY > layerBBox.maxY) {
                    // margin[0] += range.height;
                    position = 'top';
                }
                _this.paddingController.registerPadding({
                    getBBox: function () {
                        return range;
                    },
                    position: position,
                }, 'outer');
            }
        });
        var viewRange = this.paddingController.processOuterPadding();
        return viewRange;
    };
    ViewLayer.prototype.viewRangeToRegion = function (viewRange) {
        var _a = this, width = _a.width, height = _a.height;
        var start = { x: 0, y: 0 }, end = { x: 1, y: 1 };
        start.x = viewRange.minX / width;
        start.y = viewRange.minY / height;
        end.x = viewRange.maxX / width;
        end.y = viewRange.maxY / height;
        return {
            start: start,
            end: end,
        };
    };
    return ViewLayer;
}(Layer));
export default ViewLayer;
