// @antv/g2plot/esm/plots/line/animation/clipIn-with-data.js
import { registerAnimation } from '../../../dependents';
import { clone, isFunction } from '@antv/util';
var plotInfo;
function clipingWithData(shape, animateCfg, cfg) {
    var geometry = shape.get('element').geometry;
    /** 动画初始状态 */
    var index = shape.get('index');
    var coord = geometry.coordinate;
    var scales = geometry.scales;
    var yScale = scales[plotInfo.options.yField];
    var shapeData = clone(shape.get('origin'));
    setClip(shape, coord);
    var clip = shape.get('clipShape');
    var parent = shape.get('parent');
    var offsetX = 12;
    var title;
    if (animateCfg.seriesField) {
        title = parent.addShape('text', {
            attrs: {
                x: coord.start.x + offsetX,
                y: 0,
                text: shapeData[0]._origin[animateCfg.seriesField],
                fill: shape.attr('stroke'),
                fontSize: 12,
                textAlign: 'start',
                textBaseline: 'middle',
            },
        });
    }
    var offsetY = title ? 16 : 0;
    var marker = parent.addShape('text', {
        attrs: {
            x: coord.start.x + offsetX,
            y: offsetY,
            text: "test" + index,
            fill: shape.attr('stroke'),
            fontSize: 12,
            textAlign: 'start',
            textBaseline: 'middle',
        },
    });
    /** 动画执行之后 */
    animateCfg.callback = function () {
        if (shape && !shape.get('destroyed')) {
            shape.setClip(null);
            clip.remove();
            marker.animate({
                opacity: 0,
            }, 300, function () {
                marker.remove();
            });
            if (title) {
                marker.animate({
                    opacity: 0,
                }, 300, function () {
                    marker.remove();
                });
            }
        }
    };
    /** 执行动画 */
    /** 准备动画参数 */
    var delay = animateCfg.delay;
    if (isFunction(delay)) {
        delay = animateCfg.delay(index);
    }
    var easing = animateCfg.easing;
    if (isFunction(easing)) {
        easing = animateCfg.easing(index);
    }
    /** 动起来 */
    var i = 0;
    clip.animate({
        width: coord.getWidth(),
    }, animateCfg.duration, easing, animateCfg.callback, delay);
    (animateCfg.onFrame = function (ratio) {
        var position = getPositionByRatio(ratio, shapeData, coord, i);
        if (!position)
            return;
        marker.attr('x', position[0] + offsetX);
        marker.attr('y', position[1] + offsetY);
        var yText = getDataByPosition(yScale, position[1], coord);
        // use formatter
        if (yScale.formatter) {
            yText = yScale.formatter(yText);
        }
        marker.attr('text', yText);
    }),
        marker.animate(animateCfg.onFrame, {
            duration: animateCfg.duration,
            easing: easing,
            callback: animateCfg.callback,
            delay: delay,
        });
    if (title) {
        title.animate({
            onFrame: function (ratio) {
                var position = getPositionByRatio(ratio, shapeData, coord, i);
                if (!position)
                    return;
                title.attr('x', position[0] + offsetX);
                title.attr('y', position[1]);
            },
        }, animateCfg.duration, easing, animateCfg.callback, delay);
    }
}
function setClip(shape, coord) {
    var start = coord.start, end = coord.end, width = coord.width, height = coord.height;
    shape.setClip({
        type: 'rect',
        attrs: {
            x: start.x,
            y: end.y,
            width: 0,
            height: height,
        },
    });
}
function getPositionByRatio(ratio, dataPoints, coord, index) {
    var data = dataPoints.data, points = dataPoints.points;
    var currentX = coord.start.x + coord.getWidth() * ratio;
    for (var i = 0; i < points.length - 1; i++) {
        var current = points[i];
        var next = points[i + 1];
        if (currentX >= current.x && currentX <= next.x) {
            var m = (next.y - current.y) / (next.x - current.x); // 斜率
            var y = current.y + m * (currentX - current.x);
            return [currentX, y];
        }
    }
}
function getDataByPosition(scale, y, coord) {
    var yRatio = (y - coord.start.y) / (coord.end.y - coord.start.y);
    return scale.invert(yRatio).toFixed(2);
}
export function getPlotOption(option) {
    plotInfo = option;
}
registerAnimation('clipingWithData', clipingWithData);
