// @opd/g2plot-react/es/.umi/umi.js
import './core/polyfill';
import { plugin } from './core/plugin';
import { createHistory } from './core/history';
import { ApplyPluginsType } from '/Users/kagawa/workspace/open-data-plan/g2plot-react/node_modules/@umijs/runtime/dist/index.js';
import { renderClient } from '/Users/kagawa/workspace/open-data-plan/g2plot-react/node_modules/@umijs/renderer-react/dist/index.js';

var getClientRender = function getClientRender() {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return plugin.applyPlugins({
    key: 'render',
    type: ApplyPluginsType.compose,
    initialValue: function initialValue() {
      return renderClient({
        // @ts-ignore
        routes: require('./core/routes').routes,
        plugin: plugin,
        history: createHistory(args.hot),
        rootElement: 'root',
        defaultTitle: 'G2Plot React'
      });
    },
    args: args
  });
};

var clientRender = getClientRender();
export default clientRender();
window.g_umi = {
  version: '3.0.9'
}; // hot module replacement
// @ts-ignore

if (module.hot) {
  // @ts-ignore
  module.hot.accept('./core/routes', function () {
    getClientRender({
      hot: true
    })();
  });
}
