// @antv/g2plot/esm/plots/stacked-column/layer.js
import { __assign, __extends } from "tslib";
import { deepMix } from '@antv/util';
import { registerPlotType } from '../../base/global';
import ConnectedArea from '../../components/connected-area';
import BaseColumnLayer from '../column/layer';
import StackColumnLabel from './component/label';
var StackedColumnLayer = /** @class */ (function (_super) {
    __extends(StackedColumnLayer, _super);
    function StackedColumnLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'stackedColumn';
        return _this;
    }
    StackedColumnLayer.getDefaultOptions = function () {
        return deepMix({}, _super.getDefaultOptions.call(this), {
            legend: {
                visible: true,
                position: 'right-top',
            },
            label: {
                visible: false,
                position: 'middle',
                offset: 0,
                adjustColor: true,
            },
            connectedArea: {
                visible: false,
                triggerOn: 'mouseenter',
            },
        });
    };
    StackedColumnLayer.prototype.init = function () {
        if (this.options.connectedArea.visible) {
            this.options.tooltip.crosshairs = null;
        }
        _super.prototype.init.call(this);
    };
    StackedColumnLayer.prototype.afterRender = function () {
        var props = this.options;
        // 绘制区域连接组件
        if (props.connectedArea.visible) {
            this.connectedArea = new ConnectedArea(__assign({ view: this.view, field: props.stackField, animation: props.animation === false ? false : true }, props.connectedArea));
        }
        _super.prototype.afterRender.call(this);
    };
    StackedColumnLayer.prototype.adjustColumn = function (column) {
        column.adjust = [
            {
                type: 'stack',
            },
        ];
    };
    StackedColumnLayer.prototype.renderLabel = function () {
        var scales = this.config.scales;
        var yField = this.options.yField;
        var scale = scales[yField];
        if (this.options.label && this.options.label.visible) {
            var label = new StackColumnLabel(__assign({ view: this.view, plot: this, formatter: scale.formatter }, this.options.label));
            label.render();
        }
    };
    StackedColumnLayer.prototype.geometryTooltip = function () {
        this.column.tooltip = {};
        var tooltipOptions = this.options.tooltip;
        if (tooltipOptions.fields) {
            this.column.tooltip.fields = tooltipOptions.fields;
        }
        if (tooltipOptions.formatter) {
            this.column.tooltip.callback = tooltipOptions.formatter;
            if (!tooltipOptions.fields) {
                this.column.tooltip.fields = [this.options.xField, this.options.yField, this.options.stackField];
            }
        }
    };
    return StackedColumnLayer;
}(BaseColumnLayer));
export default StackedColumnLayer;
registerPlotType('stackedColumn', StackedColumnLayer);
