// @antv/g2plot/esm/plots/bar/component/label.js
import { each, deepMix, clone } from '@antv/util';
import { rgb2arr, mappingColor } from '../../../util/color';
import BBox from '../../../util/bbox';
var DEFAULT_OFFSET = 8;
var BarLabel = /** @class */ (function () {
    function BarLabel(cfg) {
        this.destroyed = false;
        this.view = cfg.view;
        this.plot = cfg.plot;
        var defaultOptions = this.getDefaultOptions();
        this.options = deepMix(defaultOptions, cfg, {});
        this.init();
    }
    BarLabel.prototype.init = function () {
        var _this = this;
        this.container = this.getGeometry().labelsContainer;
        this.view.on('beforerender', function () {
            _this.clear();
            _this.plot.canvas.draw();
        });
    };
    BarLabel.prototype.render = function () {
        var _this = this;
        var _a = this.getGeometry(), elements = _a.elements, coordinate = _a.coordinate;
        this.coord = coordinate;
        each(elements, function (ele) {
            var shape = ele.shape;
            var style = clone(_this.options.style);
            var value = _this.getValue(shape);
            var position = _this.getPosition(shape, value);
            var textAlign = _this.getTextAlign(value);
            var color = _this.getTextColor(shape);
            if (_this.options.position !== 'right' && _this.options.adjustColor && color !== 'black') {
                style.stroke = null;
            }
            var formatter = _this.options.formatter;
            var content = formatter ? formatter(value) : value;
            var label = _this.container.addShape('text', {
                attrs: deepMix({}, style, {
                    x: position.x,
                    y: position.y,
                    text: content,
                    fill: color,
                    textAlign: textAlign,
                    textBaseline: 'middle',
                }),
                name: 'label',
            });
            _this.adjustLabel(label, shape);
        });
    };
    BarLabel.prototype.clear = function () {
        if (this.container) {
            this.container.clear();
        }
    };
    BarLabel.prototype.hide = function () {
        this.container.set('visible', false);
        this.plot.canvas.draw();
    };
    BarLabel.prototype.show = function () {
        this.container.set('visible', true);
        this.plot.canvas.draw();
    };
    BarLabel.prototype.destroy = function () {
        if (this.container) {
            this.container.remove();
        }
        this.destroyed = true;
    };
    BarLabel.prototype.getBBox = function () { };
    BarLabel.prototype.getPosition = function (shape, value) {
        var bbox = this.getShapeBbox(shape);
        var minX = bbox.minX, maxX = bbox.maxX, minY = bbox.minY, height = bbox.height, width = bbox.width;
        var _a = this.options, offsetX = _a.offsetX, offsetY = _a.offsetY, position = _a.position;
        var y = minY + height / 2 + offsetY;
        var dir = value < 0 ? -1 : 1;
        var x;
        if (position === 'left') {
            var root = value > 0 ? minX : maxX;
            x = root + offsetX * dir;
        }
        else if (position === 'right') {
            x = maxX + offsetX * dir;
        }
        else {
            x = minX + width / 2 + offsetX;
        }
        return { x: x, y: y };
    };
    BarLabel.prototype.getTextColor = function (shape) {
        if (this.options.adjustColor && this.options.position !== 'right') {
            var shapeColor = shape.attr('fill');
            var shapeOpacity = shape.attr('opacity') ? shape.attr('opacity') : 1;
            var rgb = rgb2arr(shapeColor);
            var gray = Math.round(rgb[0] * 0.299 + rgb[1] * 0.587 + rgb[2] * 0.114) / shapeOpacity;
            var colorBand = [
                { from: 0, to: 85, color: 'white' },
                { from: 85, to: 170, color: '#F6F6F6' },
                { from: 170, to: 255, color: 'black' },
            ];
            var reflect = mappingColor(colorBand, gray);
            return reflect;
        }
        var defaultColor = this.options.style.fill;
        return defaultColor;
    };
    BarLabel.prototype.getTextAlign = function (value) {
        var position = this.options.position;
        var alignOptions = {
            right: 'left',
            left: 'left',
            middle: 'center',
        };
        var alignOptionsReverse = {
            right: 'right',
            left: 'right',
            middle: 'center',
        };
        if (value < 0) {
            return alignOptionsReverse[position];
        }
        return alignOptions[position];
    };
    BarLabel.prototype.getValue = function (shape) {
        var data = shape.get('origin').data;
        return data[this.plot.options.xField];
    };
    BarLabel.prototype.adjustLabel = function (label, shape) {
        if (this.options.adjustPosition && this.options.position !== 'right') {
            var labelRange = label.getBBox();
            var shapeRange = shape.getBBox();
            if (shapeRange.width <= labelRange.width) {
                var xPosition = shapeRange.maxX + this.options.offsetX;
                label.attr('x', xPosition);
                label.attr('fill', this.options.style.fill);
            }
        }
    };
    BarLabel.prototype.getDefaultOptions = function () {
        var theme = this.plot.theme;
        var labelStyle = theme.label.style;
        return {
            offsetX: DEFAULT_OFFSET,
            offsetY: 0,
            style: clone(labelStyle),
            adjustPosition: true,
        };
    };
    BarLabel.prototype.getShapeBbox = function (shape) {
        var _this = this;
        var points = [];
        each(shape.get('origin').points, function (p) {
            points.push(_this.coord.convertPoint(p));
        });
        var bbox = new BBox(points[0].x, points[2].y, Math.abs(points[1].x - points[0].x), Math.abs(points[0].y - points[2].y));
        return bbox;
    };
    BarLabel.prototype.getGeometry = function () {
        var geometries = this.view.geometries;
        var lineGeom;
        each(geometries, function (geom) {
            if (geom.type === 'interval') {
                lineGeom = geom;
            }
        });
        return lineGeom;
    };
    return BarLabel;
}());
export default BarLabel;
